src/TerraformEnvironmentStageShellHookPlugin.groovy

```diff
@@ -0,0 +1,60 @@
+import static TerraformEnvironmentStage.ALL
+import static TerraformEnvironmentStage.INIT_COMMAND
+import static TerraformEnvironmentStage.PLAN
+import static TerraformEnvironmentStage.PLAN_COMMAND
+import static TerraformEnvironmentStage.APPLY
+import static TerraformEnvironmentStage.APPLY_COMMAND
+
+class TerraformEnvironmentStageShellHookPlugin implements TerraformEnvironmentStagePlugin {
+    static hooks = [
+        (ALL): new HookPoint(ALL),
+        (INIT_COMMAND): new HookPoint(INIT_COMMAND),
+        (PLAN): new HookPoint(PLAN),
+        (PLAN_COMMAND): new HookPoint(PLAN_COMMAND),
+        (APPLY): new HookPoint(APPLY),
+        (APPLY_COMMAND): new HookPoint(APPLY_COMMAND)
+    ]
```

Comment:
Would it be easier if this here initialized as an empty map?

Reply:
On the flip side, it means that ``withHook()`` would need to null-check each of the values, and construct the new HookPoint objects if they're needed. Maybe this is backwards, but I wanted to contain all of the logic of "is this hook set" inside the HookPoint class, which led to the current implementation. Though I'm certainly open to changing it if you prefer.

Reply:
To be more specific, I think there's a lot of examples like this, where you're iterating over all the hooks because they're all non-null.  Looping can be tricky in a test - there's actually no guarantee that a PLAN_COMMAND entry will even exist in the list.

```
        @Test
        void setRunBefore() {
            TerraformEnvironmentStageShellHookPlugin.withHook(PLAN_COMMAND, 'foo bar', WhenToRun.BEFORE)
            def hooks = TerraformEnvironmentStageShellHookPlugin.hooks
            hookKeys.each {
                if (it == PLAN_COMMAND) {
                    assertTrue(hooks[it].isConfigured())
                    assertEquals(hooks[it].runBefore, 'foo bar')
                    assertNull(hooks[it].runAfterAlways)
                    assertNull(hooks[it].runAfterOnFailure)
                    assertNull(hooks[it].runAfterOnSuccess)
                } else {
                    assertFalse(hooks[it].isConfigured())
                }
            }
        }
```

I was wondering if it might be more straight forward to do something like this, which obviates the need to loop, and by passes the problem of the element not being in the list.

```
        @Test
        void setRunBefore() {
            TerraformEnvironmentStageShellHookPlugin.withHook(PLAN_COMMAND, 'foo bar', WhenToRun.BEFORE)
            def hooks = TerraformEnvironmentStageShellHookPlugin.hooks
            assertEquals(1, hooks.size())
            
            def planHooks = hooks[PLAN_COMMAND]
            assertTrue(planHooks.isConfigured())
            assertEquals(planHooks.runBefore, 'foo bar')
            assertNull(planHooks.runAfterAlways)
            assertNull(planHooks.runAfterOnFailure)
            assertNull(planHooks.runAfterOnSuccess)
        }
```

Which then suggests even more simplification with:

```
        @Test
        void setRunBefore() {
            TerraformEnvironmentStageShellHookPlugin.withHook(PLAN_COMMAND, 'foo bar', WhenToRun.BEFORE)
            def planHooks = TerraformEnvironmentStageShellHookPlugin.getHooks(PLAN_COMMAND)
            assertTrue(planHooks.isConfigured())
            assertEquals(planHooks.runBefore, 'foo bar')
            assertNull(planHooks.runAfterAlways)
            assertNull(planHooks.runAfterOnFailure)
            assertNull(planHooks.runAfterOnSuccess)
        }
```

It's a suggestion though, just based on looking at the tests, and not looking at the implementation.

Reply:
So... those alternatives are missing the intended point of the test, or half of it.

```
        @Test
        void setRunBefore() {
            TerraformEnvironmentStageShellHookPlugin.withHook(PLAN_COMMAND, 'foo bar', WhenToRun.BEFORE)
            def hooks = TerraformEnvironmentStageShellHookPlugin.hooks
            hookKeys.each {
                if (it == PLAN_COMMAND) {
                    assertTrue(hooks[it].isConfigured())
                    assertEquals(hooks[it].runBefore, 'foo bar')
                    assertNull(hooks[it].runAfterAlways)
                    assertNull(hooks[it].runAfterOnFailure)
                    assertNull(hooks[it].runAfterOnSuccess)
                } else {
                    assertFalse(hooks[it].isConfigured())
                }
            }
        }
```

Is intended to test that the plan before hook, and ONLY the plan before hook, has been set. The iteration is there to ensure that none of the other hooks, even though they're present, have been touched. If there was an easy way to serialize the complete state of ``hooks`` and test against that, then I'd do it.

The iteration here was just intended to simplify:

```
assertFalse(TerraformEnvironmentStageShellHookPlugin.getHooks(INIT_COMMAND).isConfigured())
assertFalse(TerraformEnvironmentStageShellHookPlugin.getHooks(PLAN).isConfigured())
def planHooks = TerraformEnvironmentStageShellHookPlugin.getHooks(PLAN_COMMAND)
assertTrue(planHooks.isConfigured())
assertEquals(planHooks.runBefore, 'foo bar')
assertNull(planHooks.runAfterAlways)
assertNull(planHooks.runAfterOnFailure)
assertNull(planHooks.runAfterOnSuccess)
assertFalse(TerraformEnvironmentStageShellHookPlugin.getHooks(APPLY).isConfigured())
assertFalse(TerraformEnvironmentStageShellHookPlugin.getHooks(APPLY_COMMAND).isConfigured())
```

On another note:

> there's actually no guarantee that a PLAN_COMMAND entry will even exist in the list.

Why not? Am I missing something?

Reply:
> The iteration is there to ensure that none of the other hooks, even though they're present, have been touched

That's where I was suggesting that using nulls might be helpful to simply that intent.  

```
            def hooks = TerraformEnvironmentStageShellHookPlugin.hooks
            assertEquals(1, hooks.size())
            ...
            def planHooks = hooks[PLAN_COMMAND]
            ...
```

The existence of a single non-null hook confirms that all other hooks are not set.   No worries though, it was just a suggestion. 

> Is intended to test that the plan before hook, and ONLY the plan before hook, has been set.
> On another note:
>  > there's actually no guarantee that a PLAN_COMMAND entry will even exist in the list.
>
> Why not? Am I missing something?

Apparently, I missed something, and not you.  I didn't realize until now that you created a `hookKeys` array, and are iterating over **THAT**, and not iterating over the hooks - so I'm wrong.

```
            def hooks = TerraformEnvironmentStageShellHookPlugin.hooks
       -->  hookKeys.each {
                if (it == PLAN_COMMAND) {
                    assertTrue(hooks[it].isConfigured())
                    assertEquals(hooks[it].runBefore, 'foo bar')
                    assertNull(hooks[it].runAfterAlways)
                    assertNull(hooks[it].runAfterOnFailure)
                    assertNull(hooks[it].runAfterOnSuccess)
                } else {
                    assertFalse(hooks[it].isConfigured())
                }
            }
```